tools/termux_device_info.py

```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
+"""\
+Run in Termux on Android. Prints device infos for selected device.
+
+See: https://wiki.termux.com/wiki/Termux-usb
+
+Call script in Termux like this:
+    termux-usb -r -e ./tools/termux_device_info.py /dev/bus/usb/001/002
```

Comment:
What about making this a generic device_info.py script that supports both -d [vendor]:[product] and --usb-fd [fd] modes. That would make it generally useful and would serve as an example of how other apps could provide similar command line arguments. [argparse](https://docs.python.org/library/argparse.html) would make this pretty easy to do.

Reply:
Will _unfortuately_ not work as the device path is part of the termux-api command. My script is called as callback with the open file descriptor as argument. After my script finishes, the termux-api will close the file descriptor.
_You can imagine this like python `with` or JavaScript promises/callbacks._
So my script will have nothing to do with the actual device path.
You _could_ make the script a wrapper around the termux-api (`termux-usb`) command and do some magic with argument handling to also work as the callback but not sure if this is worth the effort. (in my use case)

Reply:
Right, it was more mean to make the script generally useful outside of the termux context too.

The usage would look like:
```
device_info.py [-d [vendor]:[product]] [--usb-fd [fd]]
```

so then it could be used on termux (non-root) like this:
```
termux-usb -e 'device_info.py --usb-fd' -r /dev/bus/usb/001/002
# which runs: device_info.py --usb-fd 7
```

and with termux (root), or on other systems as:
```
device_info.py -d 1d6b:0002
device_info.py -D /dev/bus/usb/001/001
device_info.py -s 1:1
```

It _is_ actually possible to support `-D` under termux (non-root): see the `open_with_termux_usb function` in https://gist.github.com/normanr/9816a5db24dda0ce87a17da342a9684c.

I'd actually recommend that the new API should rather be more like the builtin `open` function that takes a path-like object or an integer file descriptor (like the gist's `open` function). The function doesn't even need to differentiate the type of the passed argument and can just call `open(file, mode='r+b', buffering=0)` and use `f.fileno()` to get the file descriptor to create the device handle (note that the file object _must_ be referenced so that's not closed early).

termux-packages could then provide a version of pyusb appropriately patched to convert the path-like object into a file descriptor using `open_with_termux_usb`.

Typical usage would look like:
```
  dev = usb.core.open('/dev/bus/usb/001/001')
```
or if you have a file descriptor:
```
  dev = usb.core.open(7)
```

(termux/termux-api#349 would basically do this (and more) inside of libusb, but I don't think it'll be fixed any time soon).

Reply:
Wow. That's interesting and looks promising.
For me, it's somewhat in the past but I remember looking at the socket rec/send functions in python but I was not familiar with them, so I was hesitating. (I was also often looking at other code in C that used this successfully but reimplementing in Python took much effort ...)
I was also considering writing my own Java Termux extensions. (the code is on GitHub so it might have worked but that's another rabbithole)

Currently, I'm still hoping for updates in termux-api and/or pyusb to make my life easier ...

I will have to test your code :-D

Reply:
If it takes too long, maybe an alternative solution would be a separate python package. It should be possible to override (extend?) another module I think. (I saw it for example with `faiss` and `faiss-gpu` and the installation order matters but might be solved with dependencies in `setup.py`.)
So Termux users could install this termux-callback 'hack'(?).
But generally the ability to get the device from the `fd` should be useful on any plattform.

Reply:
The improvements to the socket library are new in Python 3.3 (before that you had to use libraries like [PyXAPI](https://github.com/metricube/PyXAPI) or [python-passfd](https://github.com/NightTsarina/python-passfd)). My code currently uses `device_from_fd` from this pull request (and ideally it would just be incorporated into this pull request).

I think it would go good to support this _without_ having any termux specifics in the main pyusb branch. pyusb could read the command line for an "external open function" from an environment variable (`PYUSB_OPEN`?). The command would be passed two additional arguments: 1) the command to receive the file descriptor, and 2) the name of the file that pyusb wants to open. Then termux could just set `PYUSB_OPEN='termux-usb -r -e'`. (I've updated the gist to do it this way).

Reply:
btw, I think that `device_from_path` and the supporting functions shouldn't go into `usb.core`, but rather a new module instead. I can't think of any good names though: `usb.cli`, `usb.helper`, `usb.path`?

Typical usage could just look like this:
```python
dev = usb.helper.device_from_path('/dev/bus/usb/001/001')
```
and then the `usb.core.device_from_fd` API would become a lower-level API that `device_from_path` uses.

Reply:
Hm. Or would we want both `fd` and `path` in the same package, like `cli`.

`usb.files` might also be valid but would be confusing?
`helper` and the existing `util` seem somewhat similar. 
Do we have more functions for `path`?
`cli` is often a module (in other packages) that contains functions that work as entrypoints, like `Click`, `argparse` etc. But not sure if this is convention or I just stumbled over it so much?

Reply:
Termux-packages might work. I'm just not sure if a termux-packages patch might be too hidden? It also might not work with virtual envy depending on how it is made.

I think it would be simpler and more verbose to use a python package that just extends pyusb with the wanted functionality? With version pinning of pyusb there should be no way that something breaks unexpectedly in updates.

Reply:
I'm not sure if the `fd` and `path` functions need to be in the same package, mostly because `wrap_fd` is a pretty core function, and there's no other place to implement it but in `usb.core`. 

Good point about virtual envs, that is a good idea to not try make a termux specific package. Also termux-packages prefers to not package packages available via language specific ecosystems (eg: [my request to package more-itertools](https://github.com/termux/termux-packages/issues/5756)).

I'd like to avoid having to alter existing code to get it to work with termux - which is why I thought that adding _generic_ support to pyusb for open helpers configured via the environment variable might be acceptable. I also briefly tried proot to bind /dev/bus/usb/ to a temporary directory with a symlink to /proc/self/fd/7, but Android's selinux policy didn't like that.

If adding external opener (or just `path`) support is too hacky for pyusb (it probably is), then just adding the plain `fd` function would be extremely helpful. It would be the responsibility of the main script (or other non-pyusb library) to handle `fd` or `path` arguments on the command line (which I think is why I suggested making the device_info.py script generic, so that it could show how to support `-d vendor:product`, `-D device`, and `--usb-fd fd` as command line arguments with the library changes as were initially proposed).

Reply:
example device_info.py: https://gist.github.com/normanr/93c89ab4f1ea9b6c450dae2528eeef25

Reply:
Hmm. An `external` or `opener` module?
Too hacky probably not as the whole PyUSB is a wrapper around external USB libraries. It just has to be clear.

Good point about `fd`. But I was thinking more about your snippets with recv/send for `fd`. Or your example `usb.core.open` function that can be in the extra module.

Reply:
Actually I have a better suggestion for the environment variable: instead of it defining a command to use as an external opener it would define/redefine the function used to open the device (similar to how `PYTHONBREAKPOINT` changes the behavior of breakpoint() in [PEP 553](https://www.python.org/dev/peps/pep-0553/)).

That way usb.core could have a very simple `open_device` function that works for most use cases and it can very easily be overridden to support environments like termux. The function would take a `str` (the device path) and return `usb.core.Device`. The default implementation would just `return wrap_fd(os.open(device_path, is.O_RDRW))`.

Then the termux support can exist in another package that is not a core part of pyusb and is easy to install and enable without having to change existing scripts (although they do need to use the `open_device` function). I wonder if there's a need to support existing scripts that only use `find`.

At that point it might be a smaller change to pyusb to make the list of backends configurable with an environment variable instead. Then termux support could be added as a pyusb backend and it would support `find` and all existing scripts that use pyusb without any changes required to use a new `open_device` function. I like this option the most because it's the lease invasive and most powerful (and basically what https://github.com/termux/termux-api/issues/349 would do, except implemented as a pyusb backend instead of inside libusb1)